### ISSUE TITLE
autokey script omits last character

### ISSUE THREAD

**Issue by freddyw** [autokey/autokey#354]
## Classification:
Bug
(Pick one of: Bug, Crash/Hang/Data Loss, Performance, UI/Usability, Feature (New), Enhancement)

## Reproducibility:
Always
(Pick one of: Always, Sometimes, Rarely, Unable, I Didn't Try)

## Version

AutoKey version: 
0.95.4, 0.95.7
Used GUI (Gtk, Qt, or both): 
Gtk

Installed via: (PPA, pip3, …).
fedora DNF, OracleLinux pip3
Linux Distribution:
Fedora 31 and OracleLinux 8
## Summary

Summary of the problem.
I have a script that retrieves a passphrase from the gnome-keyring. It works flawless when I use a python script 'gkeyring'. That is a python2 script however. So I wanted to use secret-tool instead (libsecret rpm) as this is (default) part of the OS. But then I use secret-tool to extract a passphrase (which it does fine on its own) and then use autokey the last character is consistently omitted.

## Steps to Reproduce (if applicable)
- secret-tool store --label="somephrase" somephrase somekey
=> whatever
- $secret-tool lookup somephrase somekey
whatever
- create an autokey script
output = system.exec_command("secret-tool lookup somephrase somekey")
keyboard.send_keys(output + "\n")
- execute script

## Expected Results

'whatever' is returned by the autokey script

## Actual Results
```
2020-01-15 09:41:33,853 INFO - service - Matched hotkey phrase/script with prompt=False
2020-01-15 09:41:33,854 DEBUG - service - Script runner executing: Script('somescript')
2020-01-15 09:41:33,854 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2020-01-15 09:41:33,884 DEBUG - iomediator - Send via event interface
2020-01-15 09:41:33,885 DEBUG - interface - Sending string: 'whateve'
2020-01-15 09:41:33,900 DEBUG - interface - Send special key: ['<enter>']
whateve
2020-01-15 09:41:33,962 DEBUG - iomediator - Key.SHIFT released
2020-01-15 09:41:33,983 DEBUG - iomediator - Key.CONTROL released
```
## Notes
The same happens on both Fedora 31 (0.95.7, packaged version) and OracleLinux 8 (0.95.4, pip3 install).

**Comment by freddyw** [comment]
Using a simple bash script works fine by the way:

#!/bin/bash
output=$(secret-tool lookup somephrase somekey)
echo "${output}"

**Comment by luziferius** [comment]
Hmm. `system.exec_command` trims the last character, because most programs put an unwantes newline at the end.

Thus needs special handling for programs not doing this or a flag to disable this behaviour.

**Comment by BlueDrink9** [comment]
We can just check if the last character is `\n` right?

**Comment by luziferius** [comment]
> We can just check if the last character is `\n` right?

Basically yes.

https://github.com/autokey/autokey/blob/1c5e3ce8b45fbea49fd17c2b358211434605a89a/lib/autokey/scripting.py#L489-L499

The offending line has a comment why it trims a character.
Normally, all commands output a trailing `\n`. But it indeed does not make sense for a password manager to add spurious newlines at the end of passwords.

**Comment by BlueDrink9** [comment]
I've fixed it with tests. About to PR, should it be a bugfix branch off master? What should I call it?

**Comment by BlueDrink9** [comment]
Problem with applying it to master is that it doesn't have the new test setup. I'll apply to develop, PR, and you can cherry-pick just the fix (no tests) to master as a bugfix?

**Comment by luziferius** [comment]
In the past, I’ve branched bug fixes branches (named `bugfix/issue_<number>` off master, written tests in `develop` and then merged them into master and develop (and did some tricks to make sure that the tests pass before actually merging and publishing…)
That’s caused by the technical debt I created myself by rewriting the tests in `develop` instead of `master`…


> Problem with applying it to master is that it doesn't have the new test setup. I'll apply to develop, PR, and you can cherry-pick just the fix (no tests) to master as a bugfix?

That should work.

Then we’ll have this fix and https://github.com/autokey/autokey/commit/1c5e3ce8b45fbea49fd17c2b358211434605a89a for the next 0.95 release.

**Comment by BlueDrink9** [comment]
I tried merging into master after writing on develop, but ofc scripting.py has been broken up into multiple files, so the fix doesn't apply cleanly.

**Comment by luziferius** [comment]
I’ll "cherry-pick" the commit locally by copy&pasting the relevant lines.

As long as that stays "green" and develop can be merged back cleanly on 0.96 release, It’ll be fine…
https://github.com/autokey/autokey/compare/develop

**Comment by BlueDrink9** [comment]
https://github.com/BlueDrink9/autokey/commit/2f1f7819d8810f24a04458a95adc1c21734f8cf8 is the merge into master, could try cherry-picking that

**Comment by BlueDrink9** [comment]
Love a quick fix

**Comment by luziferius** [comment]
@BlueDrink9 Thank you for doing the fix, cherry picking worked flawless.

**Comment by freddyw** [comment]
Works like a charm, thanks guys!

**Comment by luziferius** [comment]
@freddyw Just released 0.95.10, which contains the fix for this issue.

**Comment by josephj11** [comment]
This patch caused a bug when the command returns no output.  See proposed fix in #462.

**Comment by BlueDrink9** [comment]
@josephj11 already fixed, PR just isn't merged. Fixed (poorly) by #486, then fixed better by #489. Just needs a few changes minor before merge. I could do them myself directly to the PR if I have permission.

**Comment by josephj11** [comment]
@BlueDrink9 You're our most senior developer. I will give you any permissions you want (if you tell me how to do it). Or I can merge any PR requests you vouch for.
